### PR TITLE
Changes to get clang 16 to compile

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

- Add new path for finding catch.  
- Don't build python documentation unless requested explicitly
- Add explicit instantiations for things that clang didn't produce symbols for

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Utilities/OptimizerHacks.cpp [hunk 1]
```diff
@@ -8,3 +8,16 @@ namespace optimizer_hacks {
 void indicate_value_can_be_changed(void* /*variable*/) {}
 }  // namespace optimizer_hacks
 #endif  /* defined(__clang__) && __clang__ < 11 */
+
+#if defined(__clang__) && __clang_major__ >= 15
+// clang v15+ fails to instantiate the following std library templates if
+// pre-compiled headers are used
+template std::__cxx11::basic_string<char, std::char_traits<char>,
```
Comment (wthrowe) [inline src/Utilities/OptimizerHacks.cpp:16]: `#include <string>` (Have you tested this with the PCH disabled?)

Might be worth putting this behind some combination of `defined(__GLIBCXX__)` and whether the PCH is enabled.

**Review by wthrowe — COMMENTED** [review]
I'll test this tomorrow.

**Comment by wthrowe** [comment]
Compiled for me on clang 16 and the tests passed (or rather, no more tests failed than before).

**Comment by kidder** [comment]
okay, pushed fixup add the include for string.  It compiles for me with both USE_PCH set to ON or OFF.  Is your comment about `__GLIBCXX__` related to using `USE_LIBCXX=ON` ?

**Comment by wthrowe** [comment]
Related, but I don't think `USE_LIBCXX=OFF` necessarily means we're not using libc++.  We're not turning it on in that case, but it might be enabled for some other reason (user-specified flags or the compiler default).  I think `__GLIBCXX__` should reliably indicate libstdc++.

**Comment by kidder** [comment]
okay, so you would like me to add `&& defined(__GLIBCXX__)`  to line 12

**Comment by wthrowe** [comment]
Yes, that sounds good.

**Review by wthrowe — APPROVED** [review]

**Comment by nilsvu** [comment]
@kidder did you need this just for AppleClang? The explicit instantiations fail to compile with clang 15 on CI:

```
../../../src/Utilities/OptimizerHacks.cpp:19:51: error: explicit instantiation of '_M_use_local_data' does not refer to a function template, variable template, member function, member class, or static data member
                           std::allocator<char>>::_M_use_local_data();
                                                  ^
../../../src/Utilities/OptimizerHacks.cpp:23:54: error: explicit instantiation of '_M_use_local_data' does not refer to a function template, variable template, member function, member class, or static data member
                           std::allocator<wchar_t>>::_M_use_local_data();
                                                     ^
```

**Comment by knelli2** [comment]
Locally with clang-16, I also found that these explicit instantiations failed to compile.

**Comment by nilsvu** [comment]
How did you fix it?

**Comment by knelli2** [comment]
Building with clang-14 ;)

But to use clang-16 I found this worked
```diff
diff --git a/src/Utilities/OptimizerHacks.cpp b/src/Utilities/OptimizerHacks.cpp
index 1732b4a118..443e2dbd11 100644
--- a/src/Utilities/OptimizerHacks.cpp
+++ b/src/Utilities/OptimizerHacks.cpp
@@ -9,7 +9,7 @@ void indicate_value_can_be_changed(void* /*variable*/) {}
 }  // namespace optimizer_hacks
 #endif  /* defined(__clang__) && __clang__ < 11 */

-#if defined(__clang__) && __clang_major__ >= 15 && defined(__GLIBCXX__)
+#if defined(__clang__) && __clang_major__ == 15 && defined(__GLIBCXX__)
 #include <string>
 // clang v15+ fails to instantiate the following std library templates if
 // pre-compiled headers are used
```

**Comment by nilsvu** [comment]
But @kidder needed this for clang 16?

**Comment by kidder** [comment]
Yes, on my linux desktop; @wthrowe also successfully compiles with clang 15 on his linux desktop with this change

**Comment by nilsvu** [comment]
Ubuntu 22.04 in the container fails 🤷‍♂️

**Comment by wthrowe** [comment]
Looks like a different c++ standard library implementation than the fix was for.  What gcc version are you using?

**Comment by nilsvu** [comment]
GCC 11.4.0

**Comment by wthrowe** [comment]
Looks like this implementation changed in gcc 12, so we need to only apply it for newer libstdc++ versions.  Unfortunately, there is apparently no macro that indicates what version you have.

**Comment by wthrowe** [comment]
Actually, apparently they finally added one in version 7: `_GLIBCXX_RELEASE`

**Pull request merged**